feat(lsp): signature help cycle binding

tiagoinaba opened the pull request:
<!--
  Thank you for contributing to Neovim!
  If this is your first time, check out https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#pull-requests-prs
  for our PR guidelines.
-->
Allows user to set the keymap for cycling through `signature_help`.

MariaSolOs commented:
This isn't a maintainable way to customize keymaps. As suggested in https://github.com/neovim/neovim/pull/31029#issuecomment-2453338904, `<Plug>` mappings feel like a cleaner solution here.

tiagoinaba commented:
Right, that does look better. Will get back to it.

tiagoinaba committed:
feat(lsp): implement <Plug> mappings for signature help cycling

Replace direct function mappings with `<Plug>` mappings for cycling
through overloaded signatures, providing better customization options
for users. This change keeps the default mapping (`<C-s>`) for cycling
if `<Plug>(nvim.lsp.ctrl-s)` is not mapped.

tiagoinaba commented:
Hey, thanks for the feedback. Here's the updated code - now using `<Plug>` mappings as suggested. This might be sufficient, but I'd appreciate confirmation or further guidance. Ready for review when you have time.

MariaSolOs commented:
I don't know enough about plug mapping to properly review this. @lewis6991/@justinmk perhaps you could take a look at this?

MariaSolOs commented on runtime/doc/lsp.txt:
```diff
@@ -1821,7 +1821,8 @@ rename({new_name}, {opts})                              *vim.lsp.buf.rename()*
     Displays signature information about the symbol under the cursor in a
-    floating window.
+    floating window. Allows cycling through signature overloads by mapping
+    `<Plug>(CycleSignature)`.
```
It would be helpful to include an example of how to configure this.

MariaSolOs commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -345,7 +345,8 @@ local sig_help_ns = api.nvim_create_namespace('nvim.lsp.signature_help')
 --- Displays signature information about the symbol under the cursor in a
---- floating window.
+--- floating window. Allows cycling through signature overloads by mapping
+--- `<Plug>(CycleSignature)`.
```
We should better scope the name here:
```suggestion
--- `<Plug>(LspCycleSignature)`.
```

tiagoinaba commented:
Included a lua example and changed the `<Plug>` name.

justinmk commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -420,7 +433,7 @@ function M.signature_help(config)
 
     if can_cycle then
-      vim.keymap.set('n', '<C-s>', function()
+      vim.keymap.set('n', '<Plug>(LspCycleSignature)', function()
```
Let's following the same "dot" convention as we do for namespaces, autocmds, etc.
```suggestion
      vim.keymap.set('n', '<Plug>(nvim.lsp.ctrl-s)', function()
```

justinmk replied:
`<C-s>` should still be the default. Otherwise everyone has to map it manually now?

IOW, map to both. Skip the default `<C-s>` mapping if a mapping to the `<Plug>` mapping exists.

MariaSolOs commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -387,7 +395,12 @@ function M.signature_help(config)
+            ' (%d/%d)%s',
+            idx,
+            total,
+            can_cycle and ' (<Plug>(LspCycleSignature) to cycle)' or ''
```
Is there a way to get the "real" left-hand side of the mapping in case the user has changed the default `<C-s>` mapping?

tiagoinaba replied:
Was thinking about that too. Thought of this, but it feels kinda bad:

```lua
    local existing_cycle_sig_mapping = nil
    for _, map in ipairs(vim.api.nvim_get_keymap('n')) do
      if map.rhs == '<Plug>(nvim.lsp.ctrl-s)' then
        existing_cycle_sig_mapping = map
      end
    end
```

Also, is there even a way to check if it's mapped at all? Regarding this:

> Skip the default <C-s> mapping if a mapping to the `<Plug>` mapping exists.

tiagoinaba commented:
hey, got caught up with work and school, but should clear up this week

tiagoinaba commented:
Hey, this is good for review, whenever you have a chance. It keeps `<C-s>` as the default.

> Was thinking about that too. Thought of this, but it feels kinda bad:
> 
> ```lua
>     local existing_cycle_sig_mapping = nil
>     for _, map in ipairs(vim.api.nvim_get_keymap('n')) do
>       if map.rhs == '<Plug>(nvim.lsp.ctrl-s)' then
>         existing_cycle_sig_mapping = map
>       end
>     end
> ```
> 
> Also, is there even a way to check if it's mapped at all? Regarding this:
> 
> > Skip the default  mapping if a mapping to the `<Plug>` mapping exists.

I did leave this in, but let me know if I should remove it.

justinmk commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -373,6 +381,13 @@ function M.signature_help(config)
     local can_cycle = total > 1 and config.focusable ~= false
     local idx = active_signature - 1
 
+    local existing_cycle_sig_mapping = nil
```
```suggestion
    local keymap = nil
```

justinmk commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -373,6 +381,13 @@ function M.signature_help(config)
     local idx = active_signature - 1
 
+    local existing_cycle_sig_mapping = nil
+    for _, map in ipairs(vim.api.nvim_get_keymap('n')) do
+      if map.rhs == '<Plug>(nvim.lsp.ctrl-s)' then
```
yes `hasmapto` is what I was thinking of, thanks zeer

justinmk replied:
iterating through mappings just to show a better description message is not a sustainable pattern. if we want to do that kind of thing, it needs to be done in a maintainable way. implementing lots of code for every case like this is not acceptable.

for now use `hasmapto` to support the basics, and can we revisit improving the UI later

tiagoinaba commented:
Hey, I don't think this has been released yet. Are you on v0.11.3?
